Adapt dose engine concept to dev_varRBErobOpt

wahln opened the pull request:
This draft PR adapts the dose engine concept to the dev_varRBErobOpt branch. Since the dose engine concept is currently residing on the dev branch, this PR prepares the union of dev and dev_varRBErobOpt.

Some adaptations are still to be done (mostly with regards to biological model management and maybe scenario management  - now would be the time to change it). 

I will assign a few reviewers to do a first rough validation of this refactoring.

- All dose calculation algorithms and MC interfaces are now engines. This means propMC should be dropped (as MonteCarlo engines now have their own properties).
- Robustness for pencil-beam engines is mainly managed in the base classes. 
- There's probably some inconsistencies with biological models.

JenHardt commented on doseCalc/+DoseEngines/matRad_PencilBeamEngineAbstract.m:
```diff
@@ -0,0 +1,630 @@
+            % limit rotated coordinates to positions where ray tracing is availabe
+            %radDepthsMat = cellfun(@(radDepthCube) matRad_interp3(dij.ctGrid.x,  dij.ctGrid.y,   dij.ctGrid.z,radDepthCube,dij.doseGrid.x,dij.doseGrid.y',dij.doseGrid.z,'nearest'),radDepthsMat,'UniformOutput',false);
+            
+            coordIsValid = ~isnan(radDepthVdoseGrid{1});
```
the rad depth are not considered correctly here, ithink for 4D it only take the raddepth from the 1 ct scen, not sure right now what the best fix is

wahln commented:
There seems to be a bigger issue with scenario management in the pencil-beam engines. Will investigate and update.

wahln commented on doseCalc/+DoseEngines/matRad_PencilBeamEngineAbstract.m:
```diff
@@ -0,0 +1,630 @@
+                this.tmpMatrixContainers.(names{n}) = cell(this.numOfBixelsContainer,this.multScen.numOfCtScen,this.multScen.totNumShiftScen,this.multScen.totNumRangeScen);
+                for ctScen = 1:this.multScen.numOfCtScen
+                    for shiftScen = 1:this.multScen.totNumShiftScen
+                        for rangeShiftScen = 1:this.multScen.totNumRangeScen
```
I streamlined this quite a bit

wahln commented on doseCalc/+DoseEngines/matRad_PencilBeamEngineAbstract.m:
```diff
@@ -0,0 +1,630 @@
+                for i = 1:dij.numOfBeams % loop over all beams
+
+                    %Initialize Beam Geometry
+                    currBeam = this.initBeam(dij,ct,cst,stf,i);
```
should be fixed

remocristoforetti commented on doseCalc/+DoseEngines/matRad_ParticlePencilBeamEngineAbstract.m:
```diff
@@ -0,0 +1,838 @@
+            subSelectIx = currBeam.radDepths{1} < (this.machine.data(maxEnergyIx).depths(end) - radDepthOffset);
+            currBeam.ixRadDepths = currBeam.ixRadDepths(subSelectIx);
+            currBeam.subIxVdoseGrid = currBeam.subIxVdoseGrid(subSelectIx);
+            currBeam.radDepths = cellfun(@(rd) rd(subSelectIx),currBeam.radDepths,'UniformOutput',false);
```
I think here currBeam.geoDepths is not filtered by subSelectIx and is kept in full length.
After in `matRad_ParticlePencilBeamEngineAbstract.getRayGeometryFromBeam()`  geoDepths and radDepths are arrays with different number of elements but the same indexes `ix` are selected on the two arrays.

wahln replied:
Yes, there is some inconsistency between geodepths and raddepths.

remocristoforetti commented on doseCalc/+DoseEngines/matRad_ParticlePencilBeamEngineAbstract.m:
```diff
@@ -0,0 +1,838 @@
+            ray.geoDepths = currBeam.geoDepths{1}(ix);
+            ray.radDepths = currBeam.radDepths{1}(ix);
+            ray.ix = currBeam.ixRadDepths(ix);
+            ray.subIxVdoseGrid = currBeam.subIxVdoseGrid(ix);
```
this line is the only additional line wrt superclass function `matRad_PencilBeamEngineAbstract.getRayGeometryFromBeam()`. Other lines are identical, could simply call the superclass function with `getRayGeometryFromBeam@superclassName`

wahln replied:
I think it is not so easy. The problem are the return options of the calcGeoDists function. Our Particle pencil-beam algorithm does not need distances in x/y, but only the radial distance. Finesampling needs x/y lateral distances, and photons need the lateral distances projected onto the isocenter plane. So each of these functions uses different calls ot the geodists. but maybe there is a better way to modularize, I will investigate.

JenHardt commented on doseCalc/+DoseEngines/matRad_ParticlePencilBeamEngineAbstract.m:
```diff
@@ -0,0 +1,838 @@
+            radDepthOffset = this.machine.data(maxEnergyIx).offset + minRaShi;
+
+            % apply limit in depth
+            subSelectIx = currBeam.radDepths{1} < (this.machine.data(maxEnergyIx).depths(end) - radDepthOffset);
```
also in the other branch we do currIx = currRadDepths <= machine.data(energyIx).depths(end) + offsetRadDepth;, so the offset switches sign

remocristoforetti replied:
Between new and old branch the value of the offset also changes sign (in the simple situations I checked up to now), so it compensates and the correction is in the same direction.

wahln commented:
Reintegrated the multiple Gaussian mechanism for the pencil-beam engine

remocristoforetti commented on doseCalc/+DoseEngines/matRad_ParticlePencilBeamEngineAbstract.m:
```diff
@@ -0,0 +1,902 @@
+        
+        function dij = loadBiologicalBaseData(this,cst,dij)
+            matRad_cfg = MatRad_Config.instance();
+            if isfield(this.machine.data,'alphaX') && isfield(this.machine.data,'betaX')
```
alphaX and betaX should be in the machine only for the carbon LEM model. For proton LET based biomodels these fields cannod be ssaved in machine.

wahln replied:
This should now also be resolved as biological dose calculation has been udpated.

remocristoforetti commented on doseCalc/+DoseEngines/matRad_ParticlePencilBeamEngineAbstract.m:
```diff
@@ -0,0 +1,902 @@
+            end
+
+            % bioDose
+            if this.calcBioDose
```
I think this interpolation only makes sense if there is an alpha/beta kernel, but that is not the case for proton LET based models. If this interpolation is specific to the LEM model only we could just let the bio model class take care of it

wahln replied:
This is resolved with a temporary solution until @remocristoforetti updates the Biological model infrastructure

remocristoforetti commented on doseCalc/+DoseEngines/matRad_PencilBeamEngineAbstract.m:
```diff
@@ -0,0 +1,643 @@
+            end
+
+            % ignore densities outside of contours
+            if this.ignoreOutsideDensities
```
Here initDoseCalc modifies the ct cube but does not output the modified structure. I think changes applied to `ct` here are forgotten

wahln replied:
It now applies the changes to `this.cubeWED` which will be used later for raytracing, so changes are persistent now.

wahln commented:
I converted the PR from draft to open. I think all functionalities from before are there, including some new ones (i.e. 4D scenario management in the scenariomodels). Let me know if you're concerns have been addressed, otherwise we can dare to do the merge. Branch is also freshly updated from master & dev.

wahln commented:
@JenHardt just showed be that 4D dose calculation is flawed because we strip rad-depths only according to ct scenario 1. Will provide a fix.

remocristoforetti commented on doseCalc/+DoseEngines/matRad_ParticlePencilBeamEngineAbstract.m:
```diff
@@ -0,0 +1,932 @@
+            if this.calcBioDose
+                ray.vTissueIndex = this.vTissueIndex(ray.subIxVdoseGrid,:);
+                ray.vAlphaX = this.vAlphaX(ray.subIxVdoseGrid);
+                ray.vBetaX  = this.vBetaX(ray.subIxVdoseGrid);
```
I think there is an inconsistency here.
this.vAlphaX and the other arrays have the dimension (#VoxelsOnDoseGrid,1) while ray.subIxVdoseGrid contains indexes referred to this.VdoseGrid which is smaller and only contains the voxels on which the dose calculation is actually performed.

wahln replied:
I fixed this by masking them when they are created and tested correct assignment:
https://github.com/e0404/matRad/blob/428bbedc0acc9e1b0be3df1471bca19995f3849c/doseCalc/%2BDoseEngines/matRad_ParticlePencilBeamEngineAbstract.m#L420-L423

remocristoforetti commented on doseCalc/+DoseEngines/matRad_ParticleMCsquareEngine.m:
```diff
@@ -0,0 +1,900 @@
+                        if this.calcLET
+                            cube = this.readMhd('LET.mhd');
+                            dij.mLETDose{ctScen,shiftScen,rangeShiftScen} = absCalibrationFactorMC2 * totalWeights * ...
+                                sparse(VdoseGrid,ones(numel(VdoseGrid),1), ...
```
VdoseGrid should be this.VdoseGrid

remocristoforetti commented on doseCalc/+DoseEngines/matRad_ParticleMCsquareEngine.m:
```diff
@@ -0,0 +1,900 @@
+
+            %Matrices for LET
+            if this.calcLET
+                this.LET_MHD_Output		 = calcDoseDirect;
```
Should this be this.config.LET_MHD_output?

remocristoforetti commented on doseCalc/+DoseEngines/@matRad_DoseEngineBase/matRad_DoseEngineBase.m:
```diff
@@ -0,0 +1,309 @@
+            % copy bixel weight vector into stf struct
+            if nargin == 5
+                if sum([stf.totalNumOfBixels]) ~= numel(w)
+                    matRad_cfg.dispEerror('weighting does not match steering information');
```
dispError, typo

wahln commented:
I fixed/enabled 4D calculation. While doing this, I (and also @JenHardt independently) figured out that 4D ct scenarios are not correctly / sensibly implemented for biological calculations. Thus I updated the Optimization Code as well to handle the alpha / betas per scenario correctly.

There is still a problem in the accumulation.

JenHardt commented on 4D/matRad_calc4dDose.m:
I just realized this implantation, will fail if dose grid and ct grid are not the same I suggest, to use the [ax,bx] = matRad_getPhotonLQMParameters(cst,numel(resultGUI.physicalDose));, and since , the dose is then already accumulated to the first ctscen, we can just use the ax and bx values from this scen

JenHardt commented on optimization/projections/matRad_VariableRBEProjection.m:
I think here this should be changed to [ctScen,~]= ind2sub(size(dij.physicalDose),scen);, since ctScenarios are in the first index otherwise it gives back the linear index

remocristoforetti commented on matRad_calcDoseInfluence.m:
```diff
@@ -0,0 +1,50 @@
+% for the given pln struct and called the associated dose calculation funtion
+%
+% call
+%   dij =  matRad_calcDoseInfluence(ct,stf,pln,cst)
```
Minor comment: order of the inputs to be adjusted, when using 'help matRad_calcDoseInfluence'
